### ISSUE TITLE
Authentication cookie remains valid even after password change - potential security issue

### ISSUE THREAD

**Issue by neTp9c** [OrchardCMS/Orchard#6014]
When a user changes own password the authorize cookies that were created earlier still work.

Two critical situation:
1. User1 logged in at computer1 (and forgot to log out). Then he logged in at computer2 and changed his password. But someone still have access to a site from the computer1 through user1 account.
2. A man1 who was a manager has left a company and his account be transferred to man2. And changing password cannot restrict an access to a site from a personal man1 computer.

**Comment by sebastienros** [comment]
We could improve this by storing some proof with the cookie. For instance a password hash, if not THE password hash itself.

**Comment by DanielStolt** [comment]
I strongly object to this change, for two reasons.

Firstly, I don't see how this is a security issue to be honest. This is how cookie authentication works in general - the cookie is issued at authentication time, and regardless of what happens with your ability to authenticate again, the cookie remains value. This is what the cookie expiration time is meant to address - it is normally configured with a time during which it's deemed acceptable that an already authenticated user retains access to the system, even if her credentials are subsequently revoked. That's why the cookie time is not "forever".

Secondly, this assumes that authentication is done by means of a password, which is not always the case. For example, we have a client where authentication can be performed in five different ways, some involving a password stored in Orchard, some involving smart cards, some involving SAML-based SSO, etc. All of them result in an authentication cookie at the Orchard level after the authentication is successfully completed.

**Comment by Piedone** [comment]
@DanielStolt if a user can remain logged in and potentially carry out tasks that only an authenticated user can (without further authorization) after their access is revoked it is a security issue. If the authentication cookie's expiration is meaningful to allow the "remember me" functionality (even if it's just a day) this is potentially a hazard. Granted this needs that the affected features don't employ any other authorization (i.e. `IAuthorizer.Authorize()`), just check for an authenticated user, which is generally a bad practice.

I agree that if we do something about this and implement a solution then it shouldn't depend on any particular means of authentication.

**Comment by sebastienros** [comment]
The security guy said to use another column in the database to stamp the cookie. We can even use a `Version` column which is updated automatically by NH when the record changes.

**Comment by DanielStolt** [comment]
Let's try to separate out the concepts here, because I fear this discussion is mixing concerns.

**Authentication** happens when user X comes to an Orchard site and presents claims, Orchard accepts those claims and issues a cookie and says "for a specific period of time starting now, ownership of this cookie means I consider you to be user X **regardless of what happens to your claims or whatever credentials you used to get those claims, wherever you did so**".

**Authorization** happens when user X comes to the site with a cookie and tries to perform some operation, Orchard checks the permissions of the user and its roles, and either allows or disallows the operation.

Once the authentication part is done, it does not (and should not) matter what happens to the user's credentials. For the purposes of the current session, the credentials are no longer relevant. Once again, that's what the cookie expiration is for. A change of password does not, and should not, mean the user's access to the system is revoked - it only means that, in order to re-authenticate, she needs to use a different password.

In some cases, the credentials are stored in the same system to which they are granting access (Orchard), such as when a user logs in with a username and password stored in the Orchard database. In such cases, we could theoretically mix one thing with the other, and invalidate a cookie if the password changes. But this is extremely arbitrary - in other cases (such as OAuth) the user will present another type of claim (an SSO ticket) based on an authentication made elsewhere, such as Facebook. Do you think we could invalidate the Orchard authentication cookie when the user goes and changes her Facebook password? :) No, of course not. This illustrates the basic principle that the **initial authentication of credentials** and the **time-limited session resulting from that initial authentication** are two separate concepts that should not be mixed.

**Comment by Piedone** [comment]
I used the terms authorization and authentication properly and meant them as such. I however, mixed in that users can get their access removed (i.e. the user account disabled) which should prevent even authentication. Sorry about that, OP's second example got me mixed up, though the issue is not about that.

So let me rephrase: disabling a user account and potentially also changing a user's password (talking about the built-in user store now; and keep in mind that the password can be changed for users by admins too, they're not just changing it for themselves) are both cases where one (may) want the users to re-authenticate with the system, regardless of the state of their authentication cookie.

E.g.: user account is disabled? You are not logged in any more (and you won't be able to log in). Password changed? You are not logged in any more. Also keep in mind that the user can be logged in on multiple clients at once: e.g. I log in on a public computer and tick "remember me" (silly me) then forget to log out. I go home and realize my mistake, so I change my password. However other people at that public computer can still happily do almost anything with my user account (without needing to know my password) and I can do nothing about it (unless I'm an admin because then I can disable the user and thus prevent it from *authorizing* against any permission, but still remain *authenticated*) for 30 days (by default) until the cookie expires.

Long story short, although I'm not the one opening this issue I think that there is a need to be able to remotely log users out.

**Comment by DanielStolt** [comment]
Perhaps the main security problem here is the 30 day (!) cookie expiration. :)

That said, @Piedone I do agree that it would be desireable to remotely sign out everywhere. I just don't think **changing your password** should be the way to do it, because:

- It only works when we store passwords in Orchard (a point which I feel you are not addressing)
- It's simply logically unrelated!

Instead there could be a button for an authenticated user labelled `Sign out everywhere` which in effect sets a new cookie stamp on the `User` record as per @sebastienros's suggestion, rendering all existing cookies useless and forcing reauthentication. No need to change your password!

(Of course, this requires that the user actually have a content item record to store the cookie stamp on, which is another thing I think we should stop depending on... but that's a separate discussion.)

**Comment by kumards** [comment]
@DanielStolt if a user authenticates through Facebook, SAML, OAuth or any other software, isn't it their responsibility to validate the user credentials and pass us authentication token?

I doubt we can persist authentication cookies for such 3rd party services mentioned above. If at all, they should do it.

I also believe that invalidating or renewing the cookie at the browser is necessary on password change for Orchard managed users.

**Comment by DanielStolt** [comment]
@kumards Yes, an external service authenticates credentials, and Orchard receives a token (e.g. a SAML ticket). But this token needs to be validated by Orchard and result in an authentication cookie issued by Orchard, otherwise the user's visit would be limited to one request.

But your comment made me realize, there is of course a difference between **persistent** cookies and session cookies - this is an excellent point that I may have missed.

When authenticating using an Orchard-managed password the user gets a **persistent** cookie. It's reasonable that this cookie should be invalidated on password change. And it would be even better if it could also be invalidated **without** a password change, as per my suggestion above.

When authenticating using externally managed credentials (SSO) the user should get a **session** cookie from Orchard. And such a cookie should never be invalidated by anything, it should simply have a very short lifespan.

**Comment by nickalbrecht** [comment]
Just a trailing comment on this. I'm a firm believer that software functionality should mimic real life. The ability to invalidate an authenticated token should be possible at any time. Disabling an account, or changing the authentication details, should trigger that invalidation.

If I go to the movie theater and buy a ticket that grants me access to the movie (credentials) and then actually enter the theater, they rip off the end and grant me the stub (authentication token). This grants me privledges to come and go for the life of the stub/token without having to provide a new ticket/credentials again. However if at anytime it is deemed that I'm a security risk to the theater, I can be removed and not permitted to re-enter, invalidating my stub/token. Previously granted access based off of the credentials or authentication token shouldn't extend beyond the change in the acceptability for the credentials (account disabled) or authentication token (password change). Now in real life, you can drop your stub, someone else can pick it up and wander into the theater, but this is a shortcoming of that system because of the inability to tie the stub/token to the original ticket purchaser (credentials), but with software that shortcoming does not exist. The comparison is not perfect, but its merit still stands I believe.

**Comment by sebastienros** [comment]
@nickalbrecht Disable the user account?

**Comment by DanielStolt** [comment]
@nickalbrecht and @sebastienros, as I understand the discussion we are actually all in agreement now. Summary of what should be done:

1. Add a button for an authenticated user labelled Sign out everywhere which in effect sets a new cookie stamp on the `User` record as per @sebastienros's suggestion, rendering all existing cookies useless and forcing reauthentication. No need to change your password!

2. Also do the same (set a new cookie stamp on the `User` record) when the user changes her password or an admin resets the user's password.

3. In SSO implementations, don't set a persistent authentication cookie (because the credentials live elsewhere and there's no way for Orchard to know when they are reset/revoked at the identity provider, so there's no opportunity to change the cookie stamp in Orchard).

**Comment by HermesSbicego-Laser** [comment]
Any news? Anyone is working on it?

**Comment by Piedone** [comment]
Not to my knowledge.

**Comment by MatteoPiovanelli** [comment]
I was just thinking about this.

A first step would be, I think, to make stuff more flexible for the `IAuthorizationService`. I am looking at `FormsAuthenticationService` and I think everything that is done there with the auth cookie should be moved to a dedicated service.
I created a gist (code does not compile, I think, and needs refactoring): https://gist.github.com/MatteoPiovanelli-Laser/6eca32667403bead62f8d9a989fa954f
The idea there is just to move all cookie handling away from the `IAuthorizationService` proper, and make it possible to customize it.
At this point I would add a setting for cookie expiration of persistent cookies, rather than keeping the 30 days constant.

As a second step I would add a service invoked by the `AuthCookieManager` to handle/provide/validate the UserData for the cookie. That makes it easy to stamp the cookie. Conceptually, I am not convinced that the cookie stamp should be in the `UserPartRecord`. @Piedone added the `LastPasswordChangeUtc` property there, that may work for an implementation that only invalidates the cookies on password changes, but does not work for the case @DaRosenberg proposed with the button "Sign Out Everywhere".
I have yet to think about this properly, but maybe the `UserData` could be generated as `string.Join(";", STUFF)`, where `STUFF` are the results of a bunch of providers. By default one provider would give `user.UserName.ToBase64()`, and another would give `_settings.Name`, to have the exact same string we have right now. When the string is being evaluated to extract/validate the user, each provider would take its substring. This way it would be easy to add the `LastPasswordChangeUtc` from one provider, and a specific stamp for "Sign out everywhere" from another.

Most of this could go directly in 1.10.x, because it would not change any existing interface, and then implementations of the providers adding to `UserData` could be added (The one using `LastPasswordChangeUtc` would go in dev where has been added).

Any opinions?

**Comment by MatteoPiovanelli** [comment]
I will make a PR with the thing I meant for the generation and validation of the userdata in the AuthenticationTicket, because words are not as clear as code.

**Comment by MatteoPiovanelli** [comment]
If I add a provider that invalidates the auth cookie on password change (and not change anything else), on password change the user will be automatically logged out (on all clients, but also on the one they are using to change password) as a side effect, because their extant auth cookie will now fail to validate.

I could (and would) keep that kind of behaviour, because having to login anew means we will forcibly update passwords that may be saved on the browser or on a password manager.

On the other hand, the behaviour now from the dev branch (I have not checked 1.10.x yet) is to keep the user logged in.
I could (but would rather not) replicate this behaviour by calling SignIn() when the password has been changed successfuly. The user would still be logged out from other clients, because the authentication cookie from those would not validate. However, I have no way of knowing if when the user last logged in they had checked the "Remember me" flag (unless I check the cookie right there, I guess).

What are your opinions?

**Comment by Piedone** [comment]
I think logging out the user on password change is fine, other services do that too. I'd just add a hint to the password change screen warning the user on what will happen.

**Comment by MatteoPiovanelli** [comment]
I think right now #7945 does (or at least allows) what's been discussed in this thread.
Feel free to pitch in and review the code there.

**Comment by BenedekFarkas** [comment]
@sebastienros I changed the milestone for this issue, because the associated PR was merged into dev (not 1.10.x), so it's not included in the 1.10.3 release.
